src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+              blockData,
+              getBroadcastId,
+              customBlockDataMap,
+              parent: block,
+              stage,
+              target
+            });
+          }
+        }
+
+        if (obscuringBlockValue) {
+          resultInputs[key] = [BIS.INPUT_DIFF_BLOCK_SHADOW, obscuringBlockValue, shadowValue];
+        } else {
+          resultInputs[key] = [BIS.INPUT_SAME_BLOCK_SHADOW, shadowValue];
+        }
+      }
+    }
+
+    return resultInputs;
+  }
+
+  interface SerializeInputsOptions {
+    stage: Stage;
+    target: Target;
+
+    blockData: sb3.Target["blocks"];
+
+    getBroadcastId: GetBroadcastId;
+    customBlockDataMap: CustomBlockDataMap;
+  }
+
+  function serializeInputs(
+    block: Block,
+    options: SerializeInputsOptions
+  ): {
+    inputs: sb3.Block["inputs"];
+    fields: sb3.Block["fields"];
+    mutation?: sb3.Block["mutation"];
+  } {
+    // Serialize a block's inputs, returning the data which should be stored on
+    // the serialized block, as well as any associated blockData.
+    //
+    // This function looks more intimidating than it ought to; most of the meat
+    // here is related to serializing specific blocks whose resultant data must
+    // be generated differently than other blocks. (Custom blocks are related
+    // the main ones to blame.)
+    //
+    // serializeInputs is in charge of converting the inputs on the provided
+    // block into the structures that Scratch 3.0 expects. There are (usually)
+    // two mappings into which inputs are stored: fields and inputs. Specific
+    // details on how these are serialized is discussed in their corresponding
+    // functions (which serializeInputs defers to for most blocks), but the
+    // gist is:
+    //
+    // * Fields store actual data, while inputs refer to shadow blocks.
+    //   (Each shadow block contains a field for storing the value of that
+    //    input, though often they are serialized as a "compressed" form that
+    //    doesn't explicitly label that field. See serializeInputsToInputs.)
+    // * Reporter blocks can be placed only into inputs - not fields.
+    //   (In actuality, the input is not replaced by a block; rather, the way
+    //    it is stored is changed to refer to the ID of the placed block, and
+    //    the shadow block contianing the field value is maintained, "obscured"
+    //    but able to be recovered if the obscuring block is moved elsewhere.)
+    //
+    // Blocks may also have a "mutation" field. This is an XML attribute
+    // mapping containing data specific to a particular instance of a block
+    // that wouldn't fit on the block's input and field mappings. Specific
+    // details may vary greatly based on the opcode.
+
+    const { blockData, stage, target, getBroadcastId, customBlockDataMap } = options;
+
+    const { fields } = serializeInputsToFields(block.inputs, {
+      fieldEntries: sb3.fieldTypeMap[block.opcode],
```

Comment:
Why pass `fieldEntries` here instead of just accessing `sb3.fieldTypeMap[block.opcode]` in `serializeInputsToFields`?

Reply:
There's no reason we're required to pass it here, since this is the only call to `serializeInputsToFields` in toSb3; I kept it for consistency, because the corresponding function for inputs requires `inputEntries` be passed (it's called below with dynamically constructed contents for custom blocks).

Reply:
IMO it's a bit misleading/confusing to pass it in since it implies a "dynamic-ness" that doesn't exist. Inputs != fields anyways, so I think it's fine if the signatures are different between the two functions.

Reply:
Sure, sounds good to me. TY for the commits btw!

Reply:
On second thought, it's better to pass in `fieldEntries` than to pass in the entire `block` so that `serializeInputsToFields` can access `block.opcode`.